### ISSUE TITLE
[UX] Revamp the search indexing settings and confirmation forms

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5975]
This is a follow-up to the feature that was introduced in #4182.

When 0% of the site is indexed (remaining items is equal to the count of items on the site):
<img src="https://user-images.githubusercontent.com/842066/193283986-cf9c8c6a-da66-40e9-a1b3-07a6bc4ec463.png" width=50%>

When 100% of the site is indexed (no "remaining" items to index - only the option to reindex everything makes sense):
<img src="https://user-images.githubusercontent.com/2423362/218322422-5a662d29-31d9-4ac7-946c-d6421c4fe74e.png" width=45%>

**Comment by argiepiano** [comment]
I agree the wording is confusing. If I remember correctly, reindexing can happen in two ways:
1. Manually through that form
2. Through cron jobs

The cron jobs may or may not be able to reindex the whole site. In which case the "remaining items" are the items still remaining in the queue to be indexed after the cron job has run - meaning those will be indexed in the next cron job. So, in the form, you are given the option to reindex the whole site again, or only those items remaining in he queue.

In your first screenshot,  the site has ALL items queued, and therefore you are seeing the same number in both options (135). In the second screenshot (I assume, a different site?) you have no queued items to reindex. 

The top option will always show the total number of items in your site.

**Comment by klonos** [comment]
> The top option will always show the total number of items in your site.

Exactly 👍🏼 ...which is why it makes no sense to have 2 options when the site is fully indexed or not indexed at all.

PS: Not sure if we should tackle this here or a separate issue, but @hosef mentioned a regression: there are times when you need to wipe/invalidate the index, but don't want to trigger a reindexing immediately and leave it to cron runs instead. That was possible before, but not now. We could add either a checkbox or a secondary submit button to the confirmation form, to simply invalidate the index and leave things to the cron runs 🤔

**Comment by argiepiano** [comment]
> which is why it makes no sense to have 2 options when the site is fully indexed or not indexed at all:

I don't understand why. I have the option of either reindexing the whole site, or only the ones queued to be indexed by cron. Care to explain your reasoning why this doesn't make sense?

**Comment by argiepiano** [comment]
>  simply invalidate the index and leave things to the cron runs

OK, I think I can answer this one, since I worked on a fix similar to this that was committed come time ago. As a site admin, I want to have the option of reindexing everything at once. Leaving it to cron is not a great option sometimes, because cron will often not reindex the whole site in one call. (when there are many things to reindex). That's why there is the manual form in the first place. Cron will reindex things no matter what you do with that form. The form is just a way to manually reindex things

**Comment by klonos** [comment]
> ... Cron will reindex things no matter what you do with that form.

Fair enough 👍🏼 ...and it did sound like a separate issue anyway. This is not clear though. Perhaps we can:
- add a short help text in the settings form, to explain that unindexed content will eventually be indexed during cron runs.
- add the word "now" to the reindex button

The above would make it clear that there is a "wait for it" way, and a "do it now" way.

**Comment by argiepiano** [comment]
Just to clarify, the two options show in the form are immediate - none of them will wait for cron. So, both of the are "now". 

The difference between the options shown in the form is that the first one will reindex your whole site at once (everything), and the second will index only the items queued for the next cron indexing. If you manually choose option 2, those are removed from the queue, and cron will basically not reindex anything.

**Comment by klonos** [comment]
> Just to clarify, the two options show in the form are immediate - none of them will wait for cron. So, both of the are "now".

Yup, that's why I'm asking if we should also handle the "run with cron" here or a separate issue. Anyway, I have a couple of ideas for a better/clearer UI and options, and they currently don't include anything related to cron runs (other than clarifying things with help text in the settings form).

**Comment by kiamlaluno** [comment]
In the first screenshot form, there is not a real choice, as the items to index are always 135; in the second screenshot form, if I want to index 0 items, I just click on the *Cancel* link.

The code should handle those cases differently.

**Comment by klonos** [comment]
> The code should handle those cases differently.

Which is why I opened this issue here 😉

**Comment by kiamlaluno** [comment]
I totally agree with should be changed. 😉

**Comment by klonos** [comment]
Here's what I currently have on my local:
- site partially indexed:
  <img src="https://user-images.githubusercontent.com/2423362/218639297-0d1bf5ca-da21-45b5-b0ef-0076fb7c9a01.png" width=50%>
- site already fully indexed (the submit button is enabled when the checkbox is ticked):
  <img src="https://user-images.githubusercontent.com/2423362/218658885-9a3a0ca1-bf6a-454b-959d-60d913dfe583.png" width=50%>
- site not indexed:
  <img src="https://user-images.githubusercontent.com/2423362/218659301-aa52ff1a-b208-4f51-b04a-0726078a05a9.png" width=50%>

Also tweaked the messages returned by `search_reindex_batch_finished()`:
<img src="https://user-images.githubusercontent.com/2423362/218663653-7b2bd407-d7b8-4440-afa5-ac40156cdc4f.png" width=50%>
<img src="https://user-images.githubusercontent.com/2423362/218663761-765d4afd-4718-4457-ad88-cd051bb318ec.png" width=50%>

**Comment by argiepiano** [comment]
@klonos, thank you for the work on this. 

- I like the idea of hiding both options (and showing a special button) when the whole sites needs to be reindexed 👍🏽 
- I also like the idea of showing both options when there are some items in the queue, so that the admin can choose to reindex everything, or only the queued items 👍🏽 
- I'm not sure about using status and info messages to convey information (like in snapshots 2 and 3) without actually having first submitted a form or run a batch process

I'm wondering if there is a more effective way to convey this info (total items, total already indexed, and total in the queue), for example, something like this (provided by Search API), perhaps moving the 56% like:  25/44 (56%)

<img width="1188" alt="Screen Shot 2023-02-14 at 8 45 51 PM" src="https://user-images.githubusercontent.com/9938978/218925367-466f0ac6-7c4b-44ee-8fc9-8ff27149dbcc.png">

This would make the info really easy to understand and make the info/status system messages unnecessary as informational message when you first open the page - with those status msgs appearing only after the process or  submission.

**Comment by klonos** [comment]
@argiepiano you beat me to it 😅 ...here's a comment that I had ready to submit:
> PS: there are many inconsistencies that I would like to fix (tweak certain things in the "Indexing" fieldset in `admin/config/search/settings` based on the insight/clarifications provided by @argiepiano earlier in this thread), and also many places where strings are being repeated. On one hand, that would mean that I would have to re-scope this issue to "Revamp the indexing settings and confirmation forms". On the other hand, if I tried to keep the PR and scope smaller, it would mean that I would unnecessarily duplicate code used elsewhere and make some "sloppy" and OCD-triggering compromises 🤔 
>
> ...I mean, look how pretty, and neat things can look:

I have started making several changes in the "Indexing" fieldset of the form in `admin/config/search/settings` but I thought that if I submitted that, then I'd be rejected as scope creep. I think I'll file a separate issue and provide the PR with my UI/UX ideas there, so that we can keep the scope of this issue small(er).

**Comment by klonos** [comment]
...you know what, what the heck! Here's a PR with everything I have on my local: https://github.com/backdrop/backdrop/pull/4343

I've added a screenshot there, to not derail this issue here as I'm waiting for feed back in the changes.

> I'm not sure about using status and info messages to convey information...

Yes, this was what I wanted feedback on.

> I'm wondering if there is a more effective way to convey this info ... something like this (provided by Search API)

Yes, I have initially added such a progress bar in the confirmation form, but it looked too busy/noisy. At the same time it looked so pretty, that I thought to use it for the index status in the search settings page (before the user clicks on the "Rebuild search index" button). One of the "big", out-of-scope-y changes in my local was extending `theme_progress_bar()` with an additional setting, to allow for a non-animated progress bar, since the animation makes sense for things like batch operations but not for "static" status (plus some CSS changes to accommodate that).
